clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -24,9 +27,135 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
                            clang::ASTContext &astctx,
                            const clang::CodeGenOptions &cgo,
                            DiagnosticsEngine &diags)
-    : astCtx(astctx), langOpts(astctx.getLangOpts()),
-      theModule{mlir::ModuleOp::create(mlir::UnknownLoc())},
-      target(astCtx.getTargetInfo()) {}
+    : builder(&context), astCtx(astctx), langOpts(astctx.getLangOpts()),
+      theModule{mlir::ModuleOp::create(mlir::UnknownLoc::get(&context))},
+      diags(diags), target(astCtx.getTargetInfo()) {}
+
+mlir::Location CIRGenModule::getLoc(SourceLocation cLoc) {
+  assert(cLoc.isValid() && "expected valid source location");
+  const SourceManager &sm = astCtx.getSourceManager();
+  PresumedLoc pLoc = sm.getPresumedLoc(cLoc);
+  StringRef filename = pLoc.getFilename();
+  return mlir::FileLineColLoc::get(builder.getStringAttr(filename),
+                                   pLoc.getLine(), pLoc.getColumn());
+}
+
+mlir::Location CIRGenModule::getLoc(SourceRange cRange) {
+  assert(cRange.isValid() && "expected a valid source range");
+  mlir::Location begin = getLoc(cRange.getBegin());
+  mlir::Location end = getLoc(cRange.getEnd());
+  mlir::Attribute metadata;
+  return mlir::FusedLoc::get({begin, end}, metadata, builder.getContext());
+}
+
+void CIRGenModule::buildGlobal(clang::GlobalDecl gd) {
+  const auto *global = cast<ValueDecl>(gd.getDecl());
+
+  if (const auto *fd = dyn_cast<FunctionDecl>(global)) {
+    // Update deferred annotations with the latest declaration if the function
+    // was already used or defined.
+    if (fd->hasAttr<AnnotateAttr>()) {
+      errorNYI(fd->getSourceRange(), "defferedAnnotations");
+    }
+    if (!fd->doesThisDeclarationHaveABody()) {
+      if (!fd->doesDeclarationForceExternallyVisibleDefinition())
+        return;
+
+      errorNYI(fd->getSourceRange(),
+               "function declaration that forces code gen");
+      return;
+    }
+  } else {
+    errorNYI(global->getSourceRange(), "global variable declaration");
+  }
+
+  // TODO(CIR): Defer emitting some global definitions until later
+  buildGlobalDefinition(gd);
+}
+
+void CIRGenModule::buildGlobalFunctionDefinition(clang::GlobalDecl gd,
+                                                 mlir::Operation *op) {
+  auto const *funcDecl = cast<FunctionDecl>(gd.getDecl());
+  auto funcOp = builder.create<mlir::cir::FuncOp>(
+      getLoc(funcDecl->getSourceRange()), funcDecl->getIdentifier()->getName());
```

Comment:
Is this using the correct name in the presence of things which rename the function? e.g., https://godbolt.org/z/E6dr4vznj

Reply:
No, I don't think it is.  Though, I know there was discussions about `alias` offline, so this is almost definitely part of the "something we'll get to in the future" patch (as this is effectively just the 'bare minimum to show we can do a thing' patch).

Reply:
Leaving that for follow-up work makes sense to me, thanks!

For these sort of questions, I think it might make sense to stub in a test case with a FIXME comment and doing the bare minimum to ensure the test doesn't crash. Then we don't accidentally forget to cover those cases again in the future when we're not looking at what was discussed in much earlier reviews. WDYT?

Reply:
There is value to that in the general case, but because of the way this is being upstreamed (and checked' against existing tests?), our existing tests + the downstream should be sufficient to make sure it isn't forgotten.

As far as 'test to make sure we don't crash'... I don't have a problem with adding tests, but I think the maturity of this project (as upstreamed), crashing is fairly well expected :) .

Reply:
I don't expect the incubator project to have tested all the weird scenarios we come up with during PR review, and I don't expect reviewers to remember test cases from earlier reviews once they become more relevant, AND I don't expect CIR folks to handle every test case up front because that likely involves doing/undoing a bunch of work and makes upstreaming harder.

How about this for a compromise: as we think of test cases during PR review, the patch author either verifies the incubator project has test coverage for that already or adds the test coverage to the incubator project, and then fixes issues wherever it makes the most sense (either as part of the PR, a separate upstream PR, or in the incubator side which then gets upstreamed in a future PR)?

Reply:
That seems completely sensible.  I might add as far as the 'fixes the issue' is "files an issue on the incubator"?  But otherwise agree.

Reply:
This is placeholder code, as Erich said the bare minimum to not choke on function definitions.  There are large amounts of code in the ClangIR incubator that will be upstreamed later to handle function definitions and their names.  Ignoring the alias issue for the moment, the name of the `cir.func` op should be the mangled name of the function, not the simple name.  But the necessary ABI classes that do the mangling haven't been upstreamed yet.

If there are existing Clang tests for edge cases (such as the `alias` attribute), then the feature won't be forgotten during ClangIR development.  Eventually every Clang test will have to be handled correctly by the ClangIR compilation path, so missing features will be noticed.  I strongly object to expecting me to create a stub test in the LLVM project; that would be a lot of effort for no real benefit.  I appreciate that that is no longer the request.  I would prefer to not worry about things like this at all.  But a compromise I would be fine with is for the PR owner to check that either (1) the feature is implemented and tested in the ClangIR incubator project, or (2) there is an issue to implement the feature in the ClangIR incubator project.

Reply:
> (1) the feature is implemented and tested in the ClangIR incubator project, or (2) there is an issue to implement the feature in the ClangIR incubator project.

I believe this is consistent with what I suggested.  We can't really depend on our existing tests, as they are... questionable in some cases, so I'd like to make sure it gets captured on the CIR project. IF it is actually already implemented though, I expect that the PR owner should write any missing test coverage.

Reply:
In Summary, so that we all can agree here, this is the chart of what I'm suggesting:

- Questioned example "Works" in the llvm/llvm-project + this patch:
  - Test coverage exists: *PR Owner* points this out
  - Test coverage does NOT exist; *PR Owner* adds requested test
- Doesn't work in llvm/llvm-project, but works in CIR Incubator:
  - Test coverage exists: *PR Owner* points this out
  - Test coverage does NOT exist: *PR owner* adds the test to the incubator.
- Doesn't work in either
  -  The *PR Owner* Files an issue against the Incubator

Reply:
That should work, with one slight change.
- Doesn't work in either
  - The _PR Owner_ points out an existing issue in the incubator or files an issue against the Incubator.

Reply:
The ClangIR incubator currently does not support the `alias` attribute as used in Aaron's example at the beginning of this thread.  I created https://github.com/llvm/clangir/issues/1048 so that feature is not forgotten.  I think that's enough to resolve this particular comment.